File: backend/app/schema/application_schema.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from uuid import UUID

class Merchant(BaseModel):
    legal_name: str
    dba_name: str
    federal_tax_id: str
    address: str
    industry: str
    annual_revenue: float

class Owner(BaseModel):
    name: str
    ssn: str
    address: str
    date_of_birth: datetime
    ownership_percentage: float

class FundingDetails(BaseModel):
    amount_requested: float
    use_of_funds: str

class Document(BaseModel):
    id: UUID
    document_type: str
    storage_url: str
    uploaded_at: datetime

class Application(BaseModel):
    id: UUID
    email_id: UUID
    received_at: datetime
    status: str
    processed_by: Optional[UUID]
    merchant: Merchant
    owners: List[Owner]
    funding_details: FundingDetails
    documents: List[Document]

# HUMAN ASSISTANCE NEEDED
# The confidence level for the Application class is 0.8, which is at the threshold.
# Please review the Application class to ensure all fields are correctly defined
# and that the relationships between classes are properly established.
# Consider adding any additional validation or methods if necessary.

File: frontend/src/schema/applicationSchema.ts
import { z } from 'zod';

const ApplicationStatus = z.enum(['PENDING', 'PROCESSING', 'APPROVED', 'REJECTED', 'REVIEW_REQUIRED']);
const DocumentType = z.enum(['ISO_APPLICATION', 'BANK_STATEMENT', 'VOIDED_CHECK']);

const MerchantSchema = z.object({
  legal_name: z.string(),
  dba_name: z.string(),
  federal_tax_id: z.string(),
  address: z.string(),
  industry: z.string(),
  annual_revenue: z.number().positive(),
});

const OwnerSchema = z.object({
  name: z.string(),
  ssn: z.string(),
  address: z.string(),
  date_of_birth: z.date(),
  ownership_percentage: z.number().min(0).max(100),
});

const FundingDetailsSchema = z.object({
  amount_requested: z.number().positive(),
  use_of_funds: z.string(),
});

const DocumentSchema = z.object({
  id: z.string().uuid(),
  document_type: DocumentType,
  storage_url: z.string().url(),
  uploaded_at: z.date(),
});

const ApplicationSchema = z.object({
  id: z.string().uuid(),
  email_id: z.string().uuid(),
  received_at: z.date(),
  status: ApplicationStatus,
  processed_by: z.string().uuid().optional(),
  merchant: MerchantSchema,
  owners: z.array(OwnerSchema),
  funding_details: FundingDetailsSchema,
  documents: z.array(DocumentSchema),
});

export {
  ApplicationSchema,
  MerchantSchema,
  OwnerSchema,
  FundingDetailsSchema,
  DocumentSchema,
  ApplicationStatus,
  DocumentType,
};